### PR TITLE
introduced l2 regularization tensorop

### PR DESCRIPTION
**Pull request by DwijayDS**

### CODE REVIEW

File: fastestimator/backend/l2_regularization.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
```
Comment (TortoiseHam) [inline fastestimator/backend/l2_regularization.py:1]:
```suggestion
# Copyright 2021 The FastEstimator Authors. All Rights Reserved.
```

File: fastestimator/backend/l2_regularization.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        l2_loss = tf.zeros(1)[0]
+        for layer in model.layers:
+            for w in layer.trainable_variables:
+                if tf.nn.l2_loss(w) != 0.0:
+                    l2_loss += tf.nn.l2_loss(w)
+
+    else:
+        raise ValueError("Unrecognized model framework: Please make sure to pass either torch or tensorflow models")
+
+    return beta * l2_loss
```
Comment (TortoiseHam) [inline fastestimator/backend/l2_regularization.py:56]:
```suggestion
    return beta * l2_loss

```

File: fastestimator/backend/l2_regularization.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+def l2_regularization(model: Union[tf.keras.Model, torch.nn.Module], beta: float = 0.01) -> Tensor:
+    """Calculate L2 Norm of model weights.
+
+    l2_reg = sum(parameter**2)/2
+
+    This method can be used with TensorFlow and Pytorch tensors
+
+    Args:
+        model: A tensorflow or pytorc model
```
Comment (TortoiseHam) [inline fastestimator/backend/l2_regularization.py:31]:
```suggestion
        model: A tensorflow or pytorch model
```

File: fastestimator/op/tensorop/loss/l2_regularization.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/l2_regularization.py:1]:
```suggestion
# Copyright 2021 The FastEstimator Authors. All Rights Reserved.
```

File: fastestimator/op/tensorop/loss/l2_regularization.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+class L2Regularizaton(LossOp):
+    """Calculate L2 Regularization Loss.
+
+    Args:
+        inputs: String key representing input loss.
+        outputs: String key under which to store the computed loss value.
+        mode: What mode(s) to execute this Op in. For example, "train", "eval", "test", or "infer". To execute
+            regardless of mode, pass None. To execute in all modes except for a particular one, you can pass an argument
+            like "!infer" or "!train".
+        model: A tensorflow or pytorc model
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/l2_regularization.py:36]:
```suggestion
        model: A tensorflow or pytorch model
```

File: fastestimator/op/tensorop/loss/l2_regularization.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                 model: Union[tf.keras.Model, torch.nn.Module],
+                 mode: Union[None, str, Iterable[str]] = None,
+                 beta: float = 0.01):
+        super().__init__(inputs=inputs, outputs=outputs, mode = mode)
+
+        self.model = model
+        self.beta = beta
+
+    def forward(self, data: Union[Tensor, List[Tensor]], state: Dict[str, Any]) -> Tensor:
+        '''
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/l2_regularization.py:52]: if there's no info here just remove this block comment

File: fastestimator/op/tensorop/loss/l2_regularization.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        outputs: String key under which to store the computed loss value.
+        mode: What mode(s) to execute this Op in. For example, "train", "eval", "test", or "infer". To execute
+            regardless of mode, pass None. To execute in all modes except for a particular one, you can pass an argument
+            like "!infer" or "!train".
+        model: A tensorflow or pytorc model
+        beta : The multiplicative factor, to weight the l2 regularization loss with the input loss
+
+    """
+    def __init__(self,
+                 inputs: Union[Tuple[str, str], List[str]],
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/l2_regularization.py:41]: what is the Tuple[str,str] input for?

File: test/PR_test/integration_test/op/tensorop/loss/test_l2_regularization.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
+# Copyright 2020 The FastEstimator Authors. All Rights Reserved.
```
Comment (TortoiseHam) [inline test/PR_test/integration_test/op/tensorop/loss/test_l2_regularization.py:1]:
```suggestion
# Copyright 2021 The FastEstimator Authors. All Rights Reserved.
```

File: test/PR_test/integration_test/op/tensorop/loss/test_l2_regularization.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        output = output.detach()
+        self.assertTrue(np.allclose(output.numpy(), 0.12751994))
+
+    def test_l2_regularization_tensorflow(self):
+        tf_l2 = fe.build(model_fn=MyNet_tf, optimizer_fn="adam")
+        l2 = L2Regularizaton(inputs='x', outputs='x', model=tf_l2)
+        output = l2.forward(data=tf.constant([0.0]), state={})
+        self.assertTrue(np.allclose(output.numpy(), 0.1275205))
+
+    def test_pytorch_weight_decay_vs_l2(self):
```
Comment (TortoiseHam) [inline test/PR_test/integration_test/op/tensorop/loss/test_l2_regularization.py:119]: we'd also want to test that the tf model gets the same (within tolerance) weights

File: fastestimator/backend/l2_regularization.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        l2_loss = torch.tensor(0.)
+        for param in model.parameters():
+            if param.requires_grad:
+                l2_loss += (torch.sum(param.pow(2))) / 2
+
+    elif isinstance(model, tf.keras.Model):
+        l2_loss = tf.zeros(1)[0]
+        for layer in model.layers:
+            for w in layer.trainable_variables:
+                if tf.nn.l2_loss(w) != 0.0:
```
Comment (TortoiseHam) [inline fastestimator/backend/l2_regularization.py:50]: is there a reason to do this if check? since we're just adding the value, what's the harm in adding an extra 0? by removing we can avoid calling l2_loss() multiple times

File: fastestimator/backend/l2_regularization.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+def l2_regularization(model: Union[tf.keras.Model, torch.nn.Module], beta: float = 0.01) -> Tensor:
+    """Calculate L2 Norm of model weights.
+
+    l2_reg = sum(parameter**2)/2
+
+    This method can be used with TensorFlow and Pytorch tensors
+
+    Args:
+        model: A tensorflow or pytorch model
+        beta : The multiplicative factor, to weight the l2 regularization loss with the input loss
```
Comment (TortoiseHam) [inline fastestimator/backend/l2_regularization.py:32]:
```suggestion
        beta: The multiplicative factor, to weight the l2 regularization loss with the input loss
```
```suggestion
        beta : The multiplicative factor, to weight the l2 regularization loss with the input loss
```

File: fastestimator/op/tensorop/loss/l2_regularization.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    """Calculate L2 Regularization Loss.
+
+    Args:
+        inputs: String key representing input loss.
+        outputs: String key under which to store the computed loss value.
+        mode: What mode(s) to execute this Op in. For example, "train", "eval", "test", or "infer". To execute
+            regardless of mode, pass None. To execute in all modes except for a particular one, you can pass an argument
+            like "!infer" or "!train".
+        model: A tensorflow or pytorch model
+        beta : The multiplicative factor, to weight the l2 regularization loss with the input loss
```
Comment (TortoiseHam) [inline fastestimator/op/tensorop/loss/l2_regularization.py:39]:
```suggestion
        beta: The multiplicative factor, to weight the l2 regularization loss with the input loss
```
```suggestion
        beta : The multiplicative factor, to weight the l2 regularization loss with the input loss
```

**Review by TortoiseHam — APPROVED** [review]
LG
